lib/pbench/agent/tool_meister_stop.py

```diff
@@ -0,0 +1,307 @@
+                # success or failure does not affect the return status of
+                # stopping the Tool Meister sub-system.
+                client.publish(group, sysinfo_path, "sysinfo", sysinfo)
+
+        # Finally we terminate the running Tool Data Sinks and the Tool
+        # Meisters, indicating if this termination is due to an interruption.
+        # Note that we always issue the
+        term_ret_val = client.terminate(group, interrupt)
+
+    # If the "end" persistent tools operation failed, we want to be sure that
+    # this command returns an error status unless running with --interrupt.
+    # "pbench-tool-meister-stop --interrupt" may occur in any TM state (e.g.,
+    # actively "running"), in which case "end" will be rejected. We won't fail
+    # the stop command because of this, since the point is to terminate as
+    # gracefully and quickly as possible.
+    #
+    # If the "end" succeeded, or if we're in interrupt mode, return the status
+    # of the terminate operation instead.
+    ret_val = end_ret_val if (end_ret_val != 0 and not interrupt) else term_ret_val
+
+    if redis_server.locally_managed():
+        # The client operations have finished, successful or unsuccessfully,
+        # and we were not given an explicit Redis server to use.  So the
+        # previous pbench-tool-meister-start must have set up the local Tool
+        # Data Sink, Tool Meister (if registered), and the Redis server.  It is
+        # our responsibility to make sure these processes shut down correctly.
+        shutdown_ret_val = graceful_shutdown(
+            benchmark_run_dir, full_hostname, group, redis_server, logger
+        )
+        if ret_val == 0:
+            # If client termination was successful, report the status of the
+            # graceful shutdown of the Tool Data Sink and the Redis server.
+            ret_val = shutdown_ret_val
+
+    return ret_val
+
+
+_NAME_ = "pbench-tool-meister-stop"
+
+
+def main():
+    parser = ArgumentParser(f"{_NAME_} [--sysinfo <list of system information items>]")
+    parser.add_argument(
+        "--sysinfo",
+        default=None,
+        help="The list of system information items to be collected.",
+    )
+    parser.add_argument(
+        "--interrupt",
+        action="store_true",
+        help="Whether or not the stop operation is in response to an interrupt.",
+    )
+    parser.add_argument(
+        "--redis-server",
+        dest="redis_server",
+        default=os.environ.get("PBENCH_REDIS_SERVER", None),
+        help=(
+            "Use an existing Redis server specified by <hostname>:<port>;"
+            " implies the use of an existing Tool Data Sink and Tool Meisters"
+            " as well."
+        ),
+    )
+    parser.add_argument(
+        "tool_group",
+        help="The tool group name of tools being run in the Tool Meisters.",
+    )
+    parsed = parser.parse_args()
+    status = start(sys.argv[0], parsed)
+    sys.exit(status)
```

Comment:
This, too, should stay behind.

Reply:
Actually, it *did* stay behind ... the problem is that it came along, too. 🍤 

Attack of the Clones.